<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>rxjs</title>
</head>
<body>
    <h1>在集合console.log来查看代码的演示效果</h1>
</body>
<script src="https://cdn.bootcdn.net/ajax/libs/rxjs/6.2.0/rxjs.umd.min.js"></script>
<script>
    //  cdn引用的直接调用
    // rxjs.interval(500).pipe(rxjs.operators.take(4)).subscribe(console.log)

    //  retryWhen(receives: (errors: Observable) => Observable, the: scheduler): Observable
    //  retryWhen处理错误时会传入类似Observable的可变对象
    const {interval} = rxjs
    const {take,map,retryWhen,delay} = rxjs.operators
    interval(1000).pipe(
        take(6),
        map(x=>{
            if(x === 4){
                throw new Error('unlucky number 4')
            }else{
                return x
            }
        }),
        //  传入报错的observable进行自定义处理
        //  延迟1秒，总共试5次
        retryWhen(err$ => err$.pipe(
            delay(1000),
            take(5)
        ))
    ).subscribe(
        x => console.log(x)
    )
</script>
</html>